src/python/pants/backend/python/rules/pex.py

```diff
@@ -64,31 +79,78 @@ class PexInterpreterConstraints:
     def __init__(self, constraints: Optional[Iterable[str]] = None) -> None:
         self.constraints = FrozenOrderedSet(sorted(constraints or ()))
 
-    def generate_pex_arg_list(self) -> List[str]:
-        args = []
-        for constraint in sorted(self.constraints):
-            args.extend(["--interpreter-constraint", constraint])
-        return args
+    @staticmethod
+    def merge_constraint_sets(constraint_sets: Iterable[Iterable[str]]) -> List[str]:
+        """Given a collection of constraints sets, merge by ORing within each individual constraint
+        set and ANDing across each distinct constraint set.
+
+        For example, given `[["CPython>=2.7", "CPython<=3"], ["CPython==3.6.*"]]`, return
+        `["CPython>=2.7,==3.6.*", "CPython<=3,==3.6.*"]`.
+        """
+        # Each element (a List[ParsedConstraint) will get ANDed.
+        parsed_constraint_sets: List[List[ParsedConstraint]] = []
+        for constraint_set in constraint_sets:
+            # Each element (a ParsedConstraint) will get ORed.
+            parsed_constraint_set: List[ParsedConstraint] = []
+            for constraint in constraint_set:
+                try:
+                    parsed_requirement = Requirement.parse(constraint)
+                except ValueError:
+                    # We allow the shorthand `>=3.7`, which gets expanded to `CPython>=3.7`. See
+                    # Pex's interpreter.py's `parse_requirement()`.
+                    parsed_requirement = Requirement.parse(f"CPython{constraint}")
+                interpreter = parsed_requirement.project_name
+                specs = set(parsed_requirement.specs)
+                parsed_constraint_set.append(ParsedConstraint(interpreter, specs))
+            parsed_constraint_sets.append(parsed_constraint_set)
+
+        def and_constraints(parsed_constraints: Sequence[ParsedConstraint]) -> ParsedConstraint:
+            merged_specs = set()
+            expected_interpreter = parsed_constraints[0][0]
+            for parsed_constraint in parsed_constraints:
+                if parsed_constraint.interpreter != expected_interpreter:
+                    attempted_interpreters = {
+                        interpreter: [
+                            str(parsed_constraint) for parsed_constraint in parsed_constraints
+                        ]
+                        for interpreter, parsed_constraints in itertools.groupby(
+                            parsed_constraints,
+                            key=lambda parsed_constraint: parsed_constraint.interpreter,
+                        )
+                    }
+                    raise ValueError(
+                        "Tried ANDing Python interpreter constraints with different interpreters. "
```

Comment:
s/interpreter/interpreter type/ in the error message. AFAIK in our lingo an interpreter is a specific instance (CPython 3.7) and an interpreter type is e.g., CPython or PyPy.